posts/improving-the-access-control-of-a-jhipster-application.md

```diff
@@ -0,0 +1,222 @@
+.antMatchers("/api/register").permitAll()
+.antMatchers("/api/activate").permitAll()
+.antMatchers("/api/authenticate").permitAll()
+.antMatchers("/api/account/reset-password/init").permitAll()
+.antMatchers("/api/account/reset-password/finish").permitAll()
+.antMatchers("/api/**").authenticated()
+.antMatchers("/management/health").permitAll()
+.antMatchers("/management/info").permitAll()
+.antMatchers("/management/**").hasAuthority(AuthoritiesConstants.ADMIN)
+```
+
+Above, the default behavior of a JHipster application with all the rules being chained using the method `antMatchers`. The pattern value `**` simply means that it will match any request so `"/api/**"` matches the application's API.
+
+More details about the methods used after the url matcher:
+- `permitAll` allows anyone
+- `authenticated` allows only authenticated users
+- `hasAuthority` allows only authenticated users with a given authority
+
+So for example, having `.antMatchers("/api/**").hasAuthority(AuthoritiesConstants.ADMIN)` will only allow **admin** to use the application's API.
+
+## Customizing an endpoint for a specific HTTP verb
+
+In some cases, the access-control will be based on the HTTP verb and it can't be done using the url patterns. In this case, the configuration must be done the method level in the resource.
+
+The code below from the class `UserResource.java` shows how JHipster does that:
+
+```java
+@GetMapping("/users/{login:" + Constants.LOGIN_REGEX + "}")
+@Timed
+public ResponseEntity<UserDTO> getUser(@PathVariable String login) {
+    log.debug("REST request to get User : {}", login);
+    return ResponseUtil.wrapOrNotFound(
+        userService.getUserWithAuthoritiesByLogin(login)
+            .map(UserDTO::new));
+}
+
+@DeleteMapping("/users/{login:" + Constants.LOGIN_REGEX + "}")
+@Timed
+@PreAuthorize("hasRole(\"" + AuthoritiesConstants.ADMIN + "\")")
+public ResponseEntity<Void> deleteUser(@PathVariable String login) {
+    log.debug("REST request to delete User: {}", login);
+    userService.deleteUser(login);
+    return ResponseEntity.ok().headers(HeaderUtil.createAlert( "userManagement.deleted", login)).build();
+}
+```
+
+Both methods use the url `/api/users/{login}` but with a different HTTP verb (`GET` and `DELETE`). The annotation `@PreAuthorize` makes sure that the user can invoke the method. In the example above, the call to delete an user will fail if the authenticated user is not an **admin**.
+
+The annotation `@PreAuthorize` can be used to have a very granular access-control and solve the problem issued in the introduction.
+
+## Entities owned by an user
+
+One last concept about the backend that might be useful to implement is to have an access-control at the user level. That way a regular user will only have access to its entities and will not be allow to view other users entities. That can be  done at the UI level but that will not protect a malicious user to brute force the API with a range of IDs for example.
```

Comment:
...have access-control...

...not be allowed to view other users' entities.

...user from brute forcing the API...